File: python/0004_Median_of_Two_Sorted_Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        a, b = (nums1, nums2) if len(nums1) < len(nums2) else (nums2, nums1)
        length = len(a) + len(b)
        half = length // 2
        left, right = 0, len(a) - 1

        while True:
            i = (left + right) // 2
            j = half - i - 2

            aLeft = a[i] if i >= 0 else float('-inf')
            aRight = a[i+1] if i + 1 < len(a) else float('inf')
            bLeft = b[j] if j >= 0 else float('-inf')
            bRight = b[j+1] if j + 1 < len(b) else float('inf')

            if aLeft <= bRight and bLeft <= aRight:
                
                if length % 2:
                    return min(aRight, bRight)
                else:
                    return (max(aLeft, bLeft) + min(aRight, bRight)) / 2
            elif aLeft > bRight:
                right = i - 1
            else:
                left = i + 1
            

File: java/0004_Median_of_Two_Sorted_Arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int[] a, b;
        
        if (nums1.length < nums2.length) {
            a = nums1;
            b = nums2;
        } else {
            a = nums2;
            b = nums1;
        }

        int length = a.length + b.length;
        int half = length / 2;
        int left = 0, right = a.length - 1;

        while (true) {
            int i = Math.floorDiv(left + right, 2);
            int j = half - i - 2;

            int aLeft = (i >= 0) ? a[i] : Integer.MIN_VALUE;
            int aRight = (i + 1 < a.length) ? a[i+1] : Integer.MAX_VALUE;
            int bLeft = (j >= 0) ? b[j] : Integer.MIN_VALUE;
            int bRight = (j + 1 < b.length) ? b[j+1] : Integer.MAX_VALUE;

            if (aLeft <= bRight && bLeft <= aRight) {
                if (length % 2 == 1) {
                    return Math.min(aRight, bRight);
                } else {
                    return (Math.max(aLeft, bLeft) + Math.min(aRight, bRight)) / 2.0;
                }
            } else if (aLeft > bRight) {
                right = i - 1;
            } else {
                left = i + 1;
            }
        }
    }
}